language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -475,93 +661,275 @@ impl GlobalRef {
         self.root.borrow().status == GlobalDataStatus::CLEAN
     }
 
-    pub fn move_from(&mut self) -> MutVal {
+    pub fn move_from(&mut self) -> VMResult<Value> {
         self.root.borrow_mut().mark_deleted();
-        self.reference.shallow_clone()
+        self.reference.copy_value()
     }
 
-    pub fn shallow_clone(&self) -> Self {
-        GlobalRef {
-            root: Rc::clone(&self.root),
-            reference: self.reference.shallow_clone(),
-        }
+    pub fn size(&self) -> AbstractMemorySize<GasCarrier> {
+        words_in(*REFERENCE_SIZE)
     }
 
-    fn size(&self) -> AbstractMemorySize<GasCarrier> {
-        *REFERENCE_SIZE
+    fn borrow_field(&self, field_offset: usize) -> VMResult<Value> {
+        let field_ref = self
+            .reference
+            .borrow_field(field_offset)?
+            .value_as::<Reference>()
+            .unwrap()
+            .0;
+        Ok(Value::global_ref(GlobalRef::new_ref(self, field_ref)))
     }
-}
 
-impl Reference for GlobalRef {
-    fn borrow_field(&self, idx: u32) -> Option<Self> {
-        match &*self.reference.peek() {
-            Value::Struct(ref vec) => match vec.get(idx as usize) {
-                Some(field_ref) => Some(GlobalRef::new_ref(self, field_ref.shallow_clone())),
-                None => None,
-            },
-            _ => None,
-        }
+    fn copy_value(&self) -> VMResult<Value> {
+        let value = self.reference.copy_value()?;
+        Ok(value)
     }
 
-    fn read_reference(self) -> MutVal {
-        self.reference.clone()
+    fn write_value(self, value: Value) {
+        self.root.borrow_mut().mark_dirty();
+        self.reference.write_value(value);
     }
 
-    fn mutate_reference(self, v: MutVal) {
-        self.root.borrow_mut().mark_dirty();
-        self.reference.mutate_reference(v);
+    fn equals(&self, v2: &GlobalRef) -> VMResult<bool> {
+        self.reference.equals(&v2.reference)
     }
 
-    fn size(&self) -> AbstractMemorySize<GasCarrier> {
-        words_in(*REFERENCE_SIZE)
+    fn equals_ref(&self, v2: &Reference) -> VMResult<bool> {
+        self.reference.equals(&v2.0)
+    }
+
+    /// Normal code should always know what type this value has. This is made available only for
+    /// tests.
+    #[allow(non_snake_case)]
+    #[doc(hidden)]
+    fn to_type_FOR_TESTING(&self) -> Type {
+        self.reference.to_type_FOR_TESTING()
     }
 }
 
-//
-// Conversion routines for the interpreter
-//
+/// API for locals in a `Frame`.
```

Comment:
At some point in the future, I think it would be interesting to look into implementing these using the stack as you described in your talk.

Reply:
agreed